Use `iter::repeat_n` to implement `Vec::extend_with`

paolobarbolini opened the pull request:
This replaces the `Vec::extend_with` manual implementation with `iter::repeat_n`, simplifying the code and fixing the issue identified in rust-lang/rust#120050.

paolobarbolini committed:
Use `iter::repeat_n` to implement `Vec::extend_with`

This replaces the manual `Vec::extend_with` implementation with
`iter::repeat_n` and `Vec::extend_trusted`. This simplifies the
code and gets LLVM to remove the special case for the last element
when `T` is trivial to clone.

joboet commented:
Let's see what effect this has in the compiler!
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 4bf70721bef3bec2c01ded28d65b0f110a8f14e6 with merge 243c5cc6905efbdf778394d96f4b83701962964f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"4bf70721bef3bec2c01ded28d65b0f110a8f14e6","merge_sha":"243c5cc6905efbdf778394d96f4b83701962964f"} -->

Noratrieb commented:
given that this branch doesn't look good without #130887, let's block it on that.
@rustbot blocked

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12096508974/job/33732346676)
Build commit: 243c5cc6905efbdf778394d96f4b83701962964f (`243c5cc6905efbdf778394d96f4b83701962964f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12096508974/job/33732346676"},"merge_sha":"243c5cc6905efbdf778394d96f4b83701962964f"} -->

rust-timer commented:
Queued 243c5cc6905efbdf778394d96f4b83701962964f with parent 76f3ff605962d7046bc1537597ceed5e12325f54, future [comparison URL](https://perf.rust-lang.org/compare.html?start=76f3ff605962d7046bc1537597ceed5e12325f54&end=243c5cc6905efbdf778394d96f4b83701962964f).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.4 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (243c5cc6905efbdf778394d96f4b83701962964f): [comparison URL](https://perf.rust-lang.org/compare.html?start=76f3ff605962d7046bc1537597ceed5e12325f54&end=243c5cc6905efbdf778394d96f4b83701962964f&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=76f3ff605962d7046bc1537597ceed5e12325f54&end=243c5cc6905efbdf778394d96f4b83701962964f&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.1%, 0.7%]   | 16    |
| Regressions ❌ <br /> (secondary)  | 0.4%  | [0.1%, 0.6%]   | 9     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-0.9%, -0.9%] | 1     |
| Improvements ✅ <br /> (secondary) | -1.3% | [-1.8%, -0.5%] | 8     |
| All ❌✅ (primary)                 | 0.3%  | [-0.9%, 0.7%]  | 17    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=76f3ff605962d7046bc1537597ceed5e12325f54&end=243c5cc6905efbdf778394d96f4b83701962964f&stat=max-rss)
<details>
<summary>Results (primary 0.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 5.6%  | [0.9%, 11.4%]   | 3     |
| Regressions ❌ <br /> (secondary)  | -     | -               | 0     |
| Improvements ✅ <br /> (primary)   | -8.1% | [-11.1%, -5.0%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -               | 0     |
| All ❌✅ (primary)                 | 0.1%  | [-11.1%, 11.4%] | 5     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=76f3ff605962d7046bc1537597ceed5e12325f54&end=243c5cc6905efbdf778394d96f4b83701962964f&stat=cycles:u)
<details>
<summary>Results (secondary -2.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.1% | [-2.1%, -2.1%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=76f3ff605962d7046bc1537597ceed5e12325f54&end=243c5cc6905efbdf778394d96f4b83701962964f&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.0%, 1.3%]   | 43    |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.2%, -0.0%] | 7     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.4%  | [-0.2%, 1.3%]  | 50    |
</details>

**Bootstrap**: 773.805s -> 774.205s (0.05%)
**Artifact size**: 332.32 MiB -> 331.64 MiB (-0.20%)

paolobarbolini commented:
Could we do another perf run? I've cherry-picked the commit from #130887 since, at least via Compiler Explorer, it looked better

joboet commented:
Sure!
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit b080e187dbc94ac64e0ee15f12be468e24704387 with merge 78d5ed40d7c459a7e5ea4cb6bea9918a493a800f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b080e187dbc94ac64e0ee15f12be468e24704387","merge_sha":"78d5ed40d7c459a7e5ea4cb6bea9918a493a800f"} -->

rust-timer commented:
Queued 78d5ed40d7c459a7e5ea4cb6bea9918a493a800f with parent 6c76ed5503966c39381fac64eb905ac45e346695, future [comparison URL](https://perf.rust-lang.org/compare.html?start=6c76ed5503966c39381fac64eb905ac45e346695&end=78d5ed40d7c459a7e5ea4cb6bea9918a493a800f).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.4 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (78d5ed40d7c459a7e5ea4cb6bea9918a493a800f): [comparison URL](https://perf.rust-lang.org/compare.html?start=6c76ed5503966c39381fac64eb905ac45e346695&end=78d5ed40d7c459a7e5ea4cb6bea9918a493a800f&stat=instructions:u).

### Overall result: ❌ regressions - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=6c76ed5503966c39381fac64eb905ac45e346695&end=78d5ed40d7c459a7e5ea4cb6bea9918a493a800f&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4% | [0.1%, 0.7%] | 17    |
| Regressions ❌ <br /> (secondary)  | 1.3% | [0.1%, 3.2%] | 16    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.4% | [0.1%, 0.7%] | 17    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6c76ed5503966c39381fac64eb905ac45e346695&end=78d5ed40d7c459a7e5ea4cb6bea9918a493a800f&stat=max-rss)
<details>
<summary>Results (primary -5.6%, secondary 0.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean   | range           | count |
|:----------------------------------:|:------:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 5.3%   | [5.3%, 5.3%]    | 1     |
| Regressions ❌ <br /> (secondary)  | 1.5%   | [1.0%, 1.8%]    | 3     |
| Improvements ✅ <br /> (primary)   | -11.1% | [-15.5%, -6.8%] | 2     |
| Improvements ✅ <br /> (secondary) | -3.2%  | [-3.2%, -3.2%]  | 1     |
| All ❌✅ (primary)                 | -5.6%  | [-15.5%, 5.3%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=6c76ed5503966c39381fac64eb905ac45e346695&end=78d5ed40d7c459a7e5ea4cb6bea9918a493a800f&stat=cycles:u)
<details>
<summary>Results (primary -1.5%, secondary 35.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.9%  | [0.9%, 0.9%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 35.1% | [33.0%, 39.1%] | 6     |
| Improvements ✅ <br /> (primary)   | -4.0% | [-4.0%, -4.0%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.5% | [-4.0%, 0.9%]  | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=6c76ed5503966c39381fac64eb905ac45e346695&end=78d5ed40d7c459a7e5ea4cb6bea9918a493a800f&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.6%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7%  | [0.1%, 1.7%]   | 33    |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.2%, -0.0%] | 9     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.6%  | [-0.2%, 1.7%]  | 42    |
</details>

**Bootstrap**: 767.632s -> 769.266s (0.21%)
**Artifact size**: 332.19 MiB -> 332.16 MiB (-0.01%)

paolobarbolini commented:
Looking at performance data borrow checking time in the Cycles benchmark for wg-grammar is the main culprit. I'm confused why that is.

The fact that this also fixes #120050 makes these changes still worth it to me.

paolobarbolini commented:
@joboet @scottmcm Do you think the combination of #130887 + this PR, plus the fact it could also fix #120050, could have a chance of working out? I'm not too sure how to interpret the benchmarks and if there's a way of fixing the wg-grammar Cycles regression.

Soveu commented:
@paolobarbolini #130887 has been merged
@rustbot -S-blocked +S-waiting-on-author

bors commented:
:hourglass: Trying commit 29565d76bb97d949162d451b45abd0870f14458b with merge c79c24ad439e510f1baa231dcaf131000f552308...
<!-- homu: {"type":"TryBuildStarted","head_sha":"29565d76bb97d949162d451b45abd0870f14458b","merge_sha":"c79c24ad439e510f1baa231dcaf131000f552308"} -->

paolobarbolini commented:
@scottmcm Do you think that these results are worth it given that this also acts as an alternative to #120050?

paolobarbolini committed:
Specialize `Vec::extend_with` for `T: TrivialClone`

After running perf on the previous commit, it was found that the
compile-time regression was too significant to ignore.
This introduces a specialized, simplified implementation for
`T: TrivialClone` that will reduce the perf regression.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

paolobarbolini commented:
I've rebased, I've changed the specialization to be `TrivialClone` instead of `Copy` (#135634) and removed `#[track_caller]` (#147042). I took `unfilled.fill_with(|| MaybeUninit::new(unsafe { ptr::read(&value) }));` from https://github.com/rust-lang/rust/blob/b4f1098e10e71139bbdc08df5ea8183e68fc0955/library/core/src/mem/maybe_uninit.rs#L1591-L1599.

@scottmcm should we do another perf run? :crossed_fingers:

rust-timer commented:
Queued f4d62e4421cd8c2012d429811d9a63c8da41587e with parent ba86c0460b0233319e01fd789a42a7276eade805, future [comparison URL](https://perf.rust-lang.org/compare.html?start=ba86c0460b0233319e01fd789a42a7276eade805&end=f4d62e4421cd8c2012d429811d9a63c8da41587e).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.3 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (f4d62e4421cd8c2012d429811d9a63c8da41587e): [comparison URL](https://perf.rust-lang.org/compare.html?start=ba86c0460b0233319e01fd789a42a7276eade805&end=f4d62e4421cd8c2012d429811d9a63c8da41587e&stat=instructions:u).

### Overall result: ❌ regressions - please read the text below


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please do so in sufficient writing along with `@rustbot label: +perf-regression-triaged`. If not, please fix the regressions and do another perf run. If its results are neutral or positive, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=ba86c0460b0233319e01fd789a42a7276eade805&end=f4d62e4421cd8c2012d429811d9a63c8da41587e&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.8% | [0.1%, 2.5%] | 8     |
| Regressions ❌ <br /> (secondary)  | 4.2% | [0.0%, 5.4%] | 7     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.8% | [0.1%, 2.5%] | 8     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ba86c0460b0233319e01fd789a42a7276eade805&end=f4d62e4421cd8c2012d429811d9a63c8da41587e&stat=max-rss)
<details>
<summary>Results (primary -0.4%, secondary -4.4%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.8%  | [2.8%, 2.8%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.6% | [-3.6%, -3.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -4.4% | [-4.4%, -4.4%] | 1     |
| All ❌✅ (primary)                 | -0.4% | [-3.6%, 2.8%]  | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ba86c0460b0233319e01fd789a42a7276eade805&end=f4d62e4421cd8c2012d429811d9a63c8da41587e&stat=cycles:u)
<details>
<summary>Results (primary 3.4%, secondary 2.9%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.4% | [3.4%, 3.4%] | 1     |
| Regressions ❌ <br /> (secondary)  | 2.9% | [2.2%, 3.6%] | 5     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 3.4% | [3.4%, 3.4%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=ba86c0460b0233319e01fd789a42a7276eade805&end=f4d62e4421cd8c2012d429811d9a63c8da41587e&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.4%, secondary 0.2%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4%  | [0.1%, 1.4%]   | 39    |
| Regressions ❌ <br /> (secondary)  | 0.2%  | [0.2%, 0.2%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.0% | [-0.0%, -0.0%] | 4     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.4%  | [-0.0%, 1.4%]  | 43    |
</details>

**Bootstrap**: 470.7s -> 472.283s (0.34%)
**Artifact size**: 388.87 MiB -> 389.00 MiB (0.03%)

scottmcm commented:
Hmm, those `eza` results are curious.  It came out *completely* differently in the codegen schedule
<img width="1750" height="1267" alt="image" src="https://github.com/user-attachments/assets/782e3729-51ee-48ee-9904-fb57e276cd17" />

scottmcm commented on library/alloc/src/vec/spec_extend_with.rs:
```diff
@@ -0,0 +1,37 @@
+        // `T::clone` for every element. Notably, `TrivialClone` also implies
+        // that the `clone` implementation will not panic, so we can avoid
+        // initialization guards and such.
+        unfilled.fill_with(|| MaybeUninit::new(unsafe { ptr::read(&value) }));
```
2. This will probably optimize away, but it's a bunch of steps for something that should be quite trivial, so I wonder if that's impacting the debug build codegen speed.

What if you skip doing anything complicated and just have
```rust
for i in 0..n {
    unsafe { ptr::write(unfilled.get_unchecked_mut(i), ptr::read(&value)) };
}
```

That also means we're not copying it *as MaybeUninit*, which will give it more information since (for example) copying `MaybeUninit<&[i32]>` is different from copying `&[i32]`.

scottmcm reviewed (commented):
I had a couple of shotgun debugging ideas -- no idea if they'll actually help, but since the perf results were still pretty red it might be good to try some stuff.

---

Aside, if I'm reading <https://doc.rust-lang.org/nightly/std/clone/trait.TrivialClone.html> right it doesn't actually say that `Drop` needs to be trivial?  So can we just not specialize away the drop guards any more, maybe?

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/133662/b4f1098e10e71139bbdc08df5ea8183e68fc0955..164c6cf48b10af19f4044c6f0cce77378c3a3716)*

rust-timer commented:
Queued 74054d7c18f8567db8284711aa3dea05a0a78d6d with parent 5b150d238fbd4fe7bc2cd3140d8e6fb4406099fa, future [comparison URL](https://perf.rust-lang.org/compare.html?start=5b150d238fbd4fe7bc2cd3140d8e6fb4406099fa&end=74054d7c18f8567db8284711aa3dea05a0a78d6d).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.7 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->
